src/aws_sigv4_internal.erl

```diff
@@ -0,0 +1,261 @@
+set_content_sha_header(#internal_signer{payload_hash = PayloadHash, options = Options}, Headers) ->
+  case PayloadHash of
+    _ when byte_size(PayloadHash) > 0, Options#v4_signer_options.add_payload_hash_header =:= true ->
+      headers_put(<<"X-Amz-Content-Sha256">>, payload_hash_string(PayloadHash), Headers);
+    _ -> Headers
+  end.
+
+-spec headers_put(binary(), binary(), headers()) -> headers().
+headers_put(Key, Val, Headers) ->
+  lists:keystore(Key, 1, Headers, {Key, Val}).
+
+-spec build_canonical_request(internal_signer()) -> {binary(), binary()}.
+build_canonical_request(Signer) ->
+  CanonMethod = build_canonical_method(Signer),
+  CanonPath = build_canonical_path(Signer),
+  CanonQuery = build_canonical_query(Signer),
+  {CanonHeaders, SignedHeaders} = build_canonical_headers(Signer),
+  {iolist_to_binary(
+     lists:join("\n",
+                [ CanonMethod
+                , CanonPath
+                , CanonQuery
+                , CanonHeaders
+                , SignedHeaders
+                , payload_hash_string(Signer#internal_signer.payload_hash)
+                ])),
+   SignedHeaders}.
+
+-spec build_canonical_method(internal_signer()) -> binary().
+build_canonical_method(Signer) ->
+  aws_sigv4_utils:toupper(Signer#internal_signer.request#request.method).
+
+-spec build_canonical_path(internal_signer()) -> binary().
+build_canonical_path(Signer) ->
+  URIMap = uri_string:parse(Signer#internal_signer.request#request.url),
+  Path = maps:get(path, URIMap, <<>>),
+  %% FIXME: we want an escaped path here, it's unclear if it already is
+  EscapedPath =
+    case byte_size(Path) =:= 0 of
+      true -> <<"/">>;
+      false -> Path
+    end,
+  case Signer#internal_signer.options#v4_signer_options.disable_double_path_escape of
+    true -> EscapedPath;
+    false -> aws_signature_utils:uri_encode_path(EscapedPath)
+  end.
+
+-spec build_canonical_query(internal_signer()) -> binary().
+build_canonical_query(Signer) ->
+  URIMap = uri_string:parse(Signer#internal_signer.request#request.url),
+  QueryString = maps:get(query, URIMap, <<"">>),
+  QueryList0 = uri_string:dissect_query(QueryString),
+  QueryMap0 =
+    lists:foldl(
+      fun({Key, Val0}, QMap) ->
+        %% treat "key" like "key="
+        Val = case Val0 of true -> <<>>; _ -> Val0 end,
+        Vals = maps:get(Key, QMap, []),
+        maps:put(Key, [Val | Vals], QMap)
+      end, maps:new(), QueryList0),
+  QueryMap1 =
+    maps:map(
+      fun(_Key, Vals) ->
+        lists:reverse(Vals)
+      end, QueryMap0),
+  QueryList2 =
+    lists:sort(
+      fun({K1, _Vs1}, {K2, _Vs2}) ->
+        K1 =< K2
+      end, maps:to_list(QueryMap1)),
+  QueryList3 =
+    lists:append(
+      lists:map(
+        fun({K, Vs}) ->
+          lists:map(fun(V) -> {K, V} end, Vs)
+        end, QueryList2)),
+  case QueryList3 of
+    [] -> <<>>;
+    _ -> binary:replace(uri_string:compose_query(QueryList3), <<"+">>, <<"%20">>)
```

Comment:
I'm a bit surprised that the manual replacing of the `+` is needed here. I kinda would have expected this to be done  `uri_string` 🤔

Reply:
I can definitely transport `+` through `parse` and `compose_query` so this replacement seems necessary.